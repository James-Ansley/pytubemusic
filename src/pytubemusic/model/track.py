"""
Data types for collections of tracks
"""
from typing import Literal

from pydantic import Field

from .base import MaybeTimedelta, Model
from .cover import MaybeCover
from .tags import TrackTags

__all__ = (
    "Track",
    "Single",
    "Split",
    "Playlist",
    "MergePlaylist",
    "Merge",
    "TrackStub",
    "AudioStub",
    "TimeStub",
)

type Track = Single | Split | Playlist | MergePlaylist | Merge


class Single(Model):
    """A single track"""

    __match_args__ = ("url", "metadata", "cover", "start", "end")

    url: str = Field(pattern=r"/watch\?v=")
    metadata: TrackTags
    cover: MaybeCover = None
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class TrackStub(Model):
    """A track that exists as a segment of a larger track"""

    __match_args__ = ("metadata", "start", "end")

    metadata: TrackTags
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class AudioStub(Model):
    """A track used only for audio with no associated metadata"""

    __match_args__ = ("url", "start", "end")

    url: str = Field(pattern=r"/watch\?v=")
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class TimeStub(Model):
    """Timestamp information"""

    __match_args__ = ("url", "start", "end")

    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class Split(Model):
    """A single track that is split into smaller individual tracks"""

    __match_args__ = ("url", "cover", "tracks")

    url: str = Field(pattern=r"/watch\?v=")
    cover: MaybeCover = None
    tracks: tuple[TrackStub, ...] = Field(min_length=1)


class Playlist(Model):
    """A playlist"""

    __match_args__ = ("url", "cover", "tracks")

    url: str = Field(pattern=r"/playlist\?list=")
    cover: MaybeCover = None
    tracks: tuple[TrackStub | Literal["DROP"], ...] | None = None


class MergePlaylist(Model):
    """Playlist tracks combined into one"""

    __match_args__ = ("url", "metadata", "cover", "tracks")

    url: str = Field(pattern=r"/playlist\?list=")
    metadata: TrackTags
    cover: MaybeCover = None
    tracks: tuple[TimeStub | Literal["DROP"], ...] | None = None


class Merge(Model):
    """Multiple tacks combined into one"""

    __match_args__ = ("metadata", "cover", "tracks")

    metadata: TrackTags
    cover: MaybeCover = None
    tracks: tuple[AudioStub, ...] = Field(min_length=1)
