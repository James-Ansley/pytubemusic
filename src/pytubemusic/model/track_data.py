from collections.abc import Iterator, Sequence
from dataclasses import dataclass
from typing import Self

from .album import Album
from .base import MaybeStr, MaybeTimedelta
from .cover import MaybeCover
from .tags import Tags
from .track import *

__all__ = ("TrackData", "AudioData", "PlaylistAudioData")

from ..utils import never


@dataclass(frozen=True)
class AudioData:
    url: str
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


@dataclass(frozen=True)
class PlaylistAudioData:
    url: str
    index: int
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


@dataclass(frozen=True)
class TrackData:
    metadata: Tags
    cover: MaybeStr
    track_parts: Sequence[AudioData | PlaylistAudioData]

    @classmethod
    def from_album(cls, album: Album) -> Iterator[Self]:
        tracks = [
            track_data
            for track in album.tracks
            for track_data in cls.from_track(track, album.metadata, album.cover)
        ]
        for i, track in enumerate(tracks, start=1):
            yield TrackData(
                metadata=track.metadata + Tags(track=i),
                cover=track.cover,
                track_parts=track.track_parts,
            )

    @classmethod
    def from_track(
          cls,
          track: Track,
          tags: Tags = Tags(),
          cover: MaybeCover = None,
    ) -> Iterator[Self]:
        match track:
            case Single():
                yield from cls.from_single(tags, cover, track)
            case Split():
                yield from cls.from_split(tags, cover, track)
            case Playlist():
                yield from cls.from_playlist(tags, cover, track)
            case MergePlaylist():
                yield from cls.from_merge_playlist(tags, cover, track)
            case Merge():
                yield from cls.from_merge(tags, cover, track)
            case _:
                never(track)

    @classmethod
    def from_playlist(
          cls,
          tags: Tags,
          cover: MaybeCover,
          track: Playlist,
    ) -> Iterator[Self]:
        cover = track.cover or cover
        for i, stub in enumerate(track.tracks):
            if stub == "DROP":
                continue
            yield cls(
                metadata=stub.metadata + tags,
                cover=cover.as_uri() if cover is not None else None,
                track_parts=(
                    PlaylistAudioData(track.url, i, stub.start, stub.end),
                )
            )

    @classmethod
    def from_split(
          cls,
          tags: Tags,
          cover: MaybeCover,
          track: Split,
    ) -> Iterator[Self]:
        cover = track.cover or cover
        for i in range(len(track.tracks) - 1):
            yield cls(
                metadata=track.tracks[i].metadata + tags,
                cover=cover.as_uri() if cover is not None else None,
                track_parts=(
                    AudioData(
                        url=track.url,
                        start=track.tracks[i].start,
                        end=track.tracks[i].end or track.tracks[i + 1].start
                    ),
                ),
            )
        last = track.tracks[-1]
        yield cls(
            metadata=last.metadata + tags,
            cover=cover.as_uri() if cover is not None else None,
            track_parts=(
                AudioData(url=track.url, start=last.start, end=last.end),
            )
        )

    @classmethod
    def from_single(
          cls,
          tags: Tags,
          cover: MaybeCover,
          track: Single,
    ) -> Iterator[Self]:
        cover = track.cover or cover
        yield cls(
            metadata=track.metadata + tags,
            cover=cover.as_uri() if cover is not None else None,
            track_parts=(
                AudioData(url=track.url, start=track.start, end=track.end),
            ),
        )

    @classmethod
    def from_merge_playlist(
          cls,
          tags: Tags,
          cover: MaybeCover,
          track: MergePlaylist
    ) -> Iterator[Self]:
        cover = track.cover or cover
        yield cls(
            metadata=track.metadata + tags,
            cover=cover.as_uri() if cover is not None else None,
            track_parts=tuple(
                PlaylistAudioData(track.url, i, stub.start, stub.end)
                for i, stub in enumerate(track.tracks)
                if stub != "DROP"
            ),
        )

    @classmethod
    def from_merge(
          cls,
          tags: Tags,
          cover: MaybeCover,
          track: Merge
    ) -> Iterator[Self]:
        cover = track.cover or cover
        yield TrackData(
            metadata=track.metadata + tags,
            cover=cover.as_uri() if cover is not None else None,
            track_parts=tuple(
                AudioData(stub.url, stub.start, stub.end)
                for stub in track.tracks
            )
        )
