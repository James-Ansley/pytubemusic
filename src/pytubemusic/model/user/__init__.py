"""
Data types for user-entered data
"""
from .album import *
from .cover import *
from .root import *
from .tags import *
from .track import *
