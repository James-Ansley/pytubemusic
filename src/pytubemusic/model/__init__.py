import pydantic as _pydantic

from .base import *
from .album import *
from .cover import *
from .tags import *
from .track import *
from .track_data import *

type RootType = "Album | Track"
RootModel = _pydantic.RootModel[RootType]
