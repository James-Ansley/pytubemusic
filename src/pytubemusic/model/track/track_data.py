"""
Normalized track data
"""
import itertools
from collections.abc import Iterator, Sequence
from dataclasses import dataclass
from typing import Self

from pytubemusic.model.types import MaybeFloat, MaybeTimedelta
from pytubemusic.model.user import (Album, Drop, MaybeCover, MediaType, Merge,
                                    MergeStub, Playlist, Single, Split, Tags,
                                    TrackStub, TrackTags, TrackType)

__all__ = ("AudioData", "PlaylistAudioData", "TrackData")


@dataclass(frozen=True)
class AudioData:
    url: str
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None

    def start_second(self) -> MaybeFloat:
        return self.start.total_seconds() if self.start is not None else None

    def duration_seconds(self) -> MaybeFloat:
        if self.start is None and self.end is not None:
            return self.end.total_seconds()
        elif self.start is not None and self.end is not None:
            return (self.end - self.start).total_seconds()
        else:
            return None


@dataclass(frozen=True)
class PlaylistAudioData:
    url: str
    index: int
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None

    def start_second(self) -> MaybeFloat:
        return self.start.total_seconds() if self.start is not None else None

    def duration_seconds(self) -> MaybeFloat:
        if self.start is None and self.end is not None:
            return self.end.total_seconds()
        elif self.start is not None and self.end is not None:
            return (self.end - self.start).total_seconds()
        else:
            return None


@dataclass(frozen=True)
class TrackData:
    metadata: Tags
    cover: MaybeCover
    parts: Sequence[AudioData | PlaylistAudioData]

    @classmethod
    def from_media(cls, media: MediaType) -> Iterator[Self]:
        match media:
            case Album():
                yield from cls.from_album(media)
            case Single() | Split() | Playlist() | Merge():
                yield from cls.from_track(media)

    @classmethod
    def from_album(cls, album: Album) -> Iterator[Self]:
        tracks = (
            track
            for album_tracks in album.tracks
            for track in cls.from_track(album_tracks)
        )
        for i, track in enumerate(tracks, start=1):
            yield TrackData(
                metadata=track.metadata + album.metadata + Tags(track=i),
                cover=track.cover or album.cover,
                parts=track.parts
            )

    @classmethod
    def from_track(cls, track: TrackType) -> Iterator[Self]:
        match track:
            case Single():
                yield from cls.from_single(track)
            case Split():
                yield from cls.from_split(track)
            case Playlist():
                yield from cls.from_playlist(track)
            case Merge():
                yield from cls.from_merge(track)

    @classmethod
    def from_single(cls, single: Single) -> Iterator[Self]:
        yield TrackData(
            metadata=Tags(**single.metadata.as_dict()),
            cover=single.cover,
            parts=[
                AudioData(url=single.url, start=single.start, end=single.end),
            ]
        )

    @classmethod
    def from_split(cls, split: Split) -> Iterator[Self]:
        end = TrackStub(metadata=TrackTags(title="unused"))
        tracks = itertools.chain(split.tracks, (end,))
        for track, next_track in itertools.pairwise(tracks):
            yield TrackData(
                metadata=Tags(**track.metadata.as_dict()),
                cover=track.cover or split.cover,
                parts=[
                    AudioData(
                        url=split.url,
                        start=track.start,
                        end=track.end or next_track.start,
                    ),
                ]
            )

    @classmethod
    def from_playlist(cls, playlist: Playlist) -> Iterator[Self]:
        i = 0
        for track in playlist.tracks:
            match track:
                case Drop():
                    i += 1
                case TrackStub():
                    yield track_data_from_track_stub(
                        track, playlist.url, i, playlist.cover
                    )
                    i += 1
                case MergeStub():
                    yield track_data_from_merge_stub(
                        track, playlist.url, i, playlist.cover,
                    )
                    i += len(track.parts)

    @classmethod
    def from_merge(cls, merge: Merge) -> Iterator[Self]:
        yield TrackData(
            metadata=Tags(**merge.metadata.as_dict()),
            cover=merge.cover,
            parts=[
                AudioData(url=part.url, start=part.start, end=part.end)
                for part in merge.parts
            ]
        )


def track_data_from_track_stub(
      track: TrackStub, url: str, i: int, playlist_cover: MaybeCover,
) -> TrackData:
    return TrackData(
        metadata=Tags(**track.metadata.as_dict()),
        cover=track.cover or playlist_cover,
        parts=[
            PlaylistAudioData(
                url=url, index=i, start=track.start, end=track.end,
            ),
        ]
    )


def track_data_from_merge_stub(
      track: MergeStub, url: str, i: int, playlist_cover: MaybeCover,
) -> TrackData:
    return TrackData(
        metadata=Tags(**track.metadata.as_dict()),
        cover=track.cover or playlist_cover,
        parts=[
            PlaylistAudioData(url=url, index=j, start=part.start, end=part.end)
            for j, part in enumerate(track.parts, start=i)
            if not isinstance(part, Drop)
        ]
    )
