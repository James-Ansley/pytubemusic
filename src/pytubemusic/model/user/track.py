"""
Data types for collections of tracks
"""
from typing import Literal

from pydantic import Field, RootModel

from pytubemusic.model.types import MaybeTimedelta
from .base import Model
from .cover import MaybeCover
from .tags import TrackTags

__all__ = (
    "Track",
    "TrackType",
    "Single",
    "Split",
    "Playlist",
    "Merge",
    "TrackStub",
    "AudioStub",
    "TimeStub",
    "MergeStub",
    "Drop",
)

type TrackType = Single | Split | Playlist | Merge


class Track:
    def __new__(cls, **kwargs) -> TrackType:
        # noinspection PyTypeChecker
        return RootModel[TrackType](**kwargs).root


class Single(Model):
    """A single track"""

    __match_args__ = ("url", "metadata", "cover", "start", "end")

    url: str = Field(pattern=r"/watch\?v=")
    metadata: TrackTags
    cover: MaybeCover = None
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class TrackStub(Model):
    """A track that exists as a segment of a larger track"""

    __match_args__ = ("metadata", "cover", "start", "end")

    metadata: TrackTags
    cover: MaybeCover = None
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class Drop(Model):
    __match_args__ = ("drop",)

    drop: Literal[True] = True


class AudioStub(Model):
    """A track used only for audio with no associated metadata"""

    __match_args__ = ("url", "start", "end")

    url: str = Field(pattern=r"/watch\?v=")
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class TimeStub(Model):
    """Timestamp information"""

    __match_args__ = ("url", "start", "end")

    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class MergeStub(Model):
    """Multiple time stamps clamped into one"""

    __match_args__ = ("metadata", "cover", "parts")

    metadata: TrackTags
    cover: MaybeCover = None
    parts: tuple[TimeStub | Drop, ...]


class Split(Model):
    """A single track that is split into smaller individual tracks"""

    __match_args__ = ("url", "cover", "tracks")

    url: str = Field(pattern=r"/watch\?v=")
    cover: MaybeCover = None
    tracks: tuple[TrackStub, ...] = Field(min_length=1)


class Playlist(Model):
    """A playlist"""

    __match_args__ = ("url", "cover", "tracks")

    url: str = Field(pattern=r"/playlist\?list=")
    cover: MaybeCover = None
    tracks: tuple[TrackStub | MergeStub | Drop, ...] = ()


class Merge(Model):
    """Multiple tacks combined into one"""

    __match_args__ = ("metadata", "cover", "tracks")

    metadata: TrackTags
    cover: MaybeCover = None
    parts: tuple[AudioStub, ...] = Field(min_length=1)
