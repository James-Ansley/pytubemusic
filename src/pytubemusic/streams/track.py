from functools import reduce
from pathlib import Path

from pytubemusic.model.audio import Audio, RawAudio
from pytubemusic.model.track import TrackData
from pytubemusic.streams.audio import fetch_audio_data
from pytubemusic.streams.images import fetch_cover_data


def fetch_track(track: TrackData, context: Path = None) -> Audio:
    return Audio(
        raw_audio=reduce(
            merge_audio,
            (fetch_audio_data(part) for part in track.parts),
        ),
        metadata=track.metadata,
        cover=fetch_cover_data(track.cover, context),
    )


def merge_audio(audio1: RawAudio, audio2: RawAudio) -> RawAudio:
    return RawAudio(
        segment=audio1.segment + audio2.segment,
        bit_rate=max(audio1.bit_rate, audio2.bit_rate),
    )
