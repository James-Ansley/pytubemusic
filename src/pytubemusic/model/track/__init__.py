from .track_data import *
