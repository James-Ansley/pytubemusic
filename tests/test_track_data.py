from datetime import timedelta
from pprint import pformat

from pytest import approx
from approvaltests import verify

from pytubemusic.model.track.track_data import (
    TrackData,
    AudioData,
    PlaylistAudioData,
)
from pytubemusic.model.user import (
    Album,
    File,
    Media,
    Merge,
    Playlist,
    Single,
    Split,
    Tags,
    Url,
)
from tests import test
from tests.utils import load_toml


@test()
def audio_data_can_list_start_seconds_and_duration_seconds():
    audio = AudioData(
        url="www.example.com/watch?v=",
        start=timedelta(seconds=1, milliseconds=500),
        end=timedelta(seconds=2, milliseconds=600),
    )
    assert audio.start_second() == approx(1.5)
    assert audio.duration_seconds() == approx(1.1)
    audio = AudioData(
        url="www.example.com/watch?v=",
        start=None,
        end=timedelta(seconds=2, milliseconds=600),
    )
    assert audio.start_second() is None
    assert audio.duration_seconds() == approx(2.6)
    audio = AudioData(
        url="www.example.com/watch?v=",
        start=timedelta(seconds=1, milliseconds=500),
        end=None
    )
    assert audio.start_second() == approx(1.5)
    assert audio.duration_seconds() is None
    audio = AudioData(
        url="www.example.com/watch?v=",
        start=None,
        end=None
    )
    assert audio.start_second() is None
    assert audio.duration_seconds() is None


@test()
def playlist_audio_data_can_list_start_seconds_and_duration_seconds():
    audio = PlaylistAudioData(
        url="www.example.com/watch?v=",
        index=0,
        start=timedelta(seconds=1, milliseconds=500),
        end=timedelta(seconds=2, milliseconds=600),
    )
    assert audio.start_second() == approx(1.5)
    assert audio.duration_seconds() == approx(1.1)
    audio = PlaylistAudioData(
        url="www.example.com/watch?v=",
        index=0,
        start=None,
        end=timedelta(seconds=2, milliseconds=600),
    )
    assert audio.start_second() is None
    assert audio.duration_seconds() == approx(2.6)
    audio = PlaylistAudioData(
        url="www.example.com/watch?v=",
        index=0,
        start=timedelta(seconds=1, milliseconds=500),
        end=None
    )
    assert audio.start_second() == approx(1.5)
    assert audio.duration_seconds() is None
    audio = PlaylistAudioData(
        url="www.example.com/watch?v=",
        index=0,
        start=None,
        end=None
    )
    assert audio.start_second() is None
    assert audio.duration_seconds() is None


@test(
    depends_on=(
          "test_track.py::singles_can_be_created_with_minimal_data",
          "test_track.py::singles_can_be_created_with_full_data",
    ),
    scope="session",
)
def singles_can_be_converted_to_track_data():
    single = Single(**load_toml("single_full.toml"))
    track_data = next(TrackData.from_single(single))
    audio = AudioData(
        url="www.example.com/watch?v=",
        start=timedelta(seconds=1, milliseconds=500),
        end=timedelta(minutes=3, seconds=21),
    )
    assert track_data.metadata == Tags(title="My Track Title")
    assert track_data.cover == File(path="data/pic.jpeg")
    assert tuple(track_data.parts) == (audio,)

    single = Single(**load_toml("single_minimal.toml"))
    track_data, = TrackData.from_single(single)
    audio = AudioData(url="www.example.com/watch?v=")
    assert track_data.metadata == Tags(title="My Track Title")
    assert track_data.cover is None
    assert tuple(track_data.parts) == (audio,)


@test(
    depends_on=(
          "test_track.py::split_tracks_can_be_created_with_minimal_data",
          "test_track.py::split_tracks_can_be_created_with_full_data",
    ),
    scope="session",
)
def split_tracks_can_be_converted_to_track_data():
    split = Split(**load_toml("split_track_full.toml"))
    track1, track2 = TrackData.from_split(split)
    audio = AudioData(
        url="www.example.com/watch?v=",
        start=timedelta(seconds=1, milliseconds=500),
        end=timedelta(seconds=12, milliseconds=300),
    )
    assert track1.metadata == Tags(title="My First Track Title")
    assert track1.cover == Url(href="www.example.com/pic2.jpeg")
    assert tuple(track1.parts) == (audio,)
    audio = AudioData(
        url="www.example.com/watch?v=",
        start=timedelta(seconds=12, milliseconds=300),
        end=timedelta(minutes=43, seconds=21),
    )
    assert track2.metadata == Tags(title="My Second Track Title")
    assert track2.cover == Url(href="www.example.com/pic1.jpeg")
    assert tuple(track2.parts) == (audio,)

    split = Split(**load_toml("split_track_minimal.toml"))
    track, = TrackData.from_split(split)
    assert track.metadata == Tags(title="My Track Title")
    assert track.cover is None
    assert tuple(track.parts) == (
        AudioData(url="www.example.com/watch?v="),
    )


@test(
    depends_on=(
          "test_track.py::playlists_can_be_created_with_minimal_data",
          "test_track.py::playlists_can_be_created_with_full_data",
          "test_track.py::playlist_tracks_can_be_merged",
    ),
    scope="session",
)
def playlists_can_be_converted_to_track_data():
    playlist = Playlist(**load_toml("playlist_full.toml"))
    track1, track2 = TrackData.from_playlist(playlist)
    assert track1.metadata == Tags(title="My Track Title")
    assert track1.cover == Url(href="www.example.com/pic2.jpeg")
    audio = PlaylistAudioData(
        url="www.example.com/playlist?list=",
        index=0,
        start=timedelta(seconds=1, milliseconds=500),
        end=timedelta(minutes=43, seconds=21),
    )
    assert tuple(track1.parts) == (audio,)

    assert track2.metadata == Tags(title="My Other Track Title")
    assert track2.cover == Url(href="www.example.com/pic1.jpeg")
    audio = PlaylistAudioData(
        url="www.example.com/playlist?list=",
        index=2,
        end=timedelta(minutes=43, seconds=21),
    )
    assert tuple(track2.parts) == (audio,)

    playlist = Playlist(**load_toml("playlist_minimal.toml"))
    assert tuple(TrackData.from_playlist(playlist)) == ()

    playlist = Playlist(**load_toml("playlist_with_merge.toml"))
    track1, track2 = TrackData.from_playlist(playlist)
    assert track1.metadata == Tags(title="My Track Title")
    assert track1.cover == Url(href="www.example.com/pic2.jpeg")
    assert tuple(track1.parts) == (
        PlaylistAudioData(
            url="www.example.com/playlist?list=",
            index=0,
            start=timedelta(seconds=1, milliseconds=500),
            end=timedelta(minutes=43, seconds=21),
        ),
        PlaylistAudioData(
            url="www.example.com/playlist?list=",
            index=2,
            start=timedelta(seconds=1),
            end=None
        ),
        PlaylistAudioData(
            url="www.example.com/playlist?list=",
            index=3,
            start=None,
            end=None
        ),
    )
    assert track2.metadata == Tags(title="My Other Track Title")
    assert track2.cover == Url(href="www.example.com/pic1.jpeg")
    audio = PlaylistAudioData(
        url="www.example.com/playlist?list=",
        index=5,
        end=timedelta(minutes=43, seconds=21),
    )
    assert tuple(track2.parts) == (audio,)


@test(
    depends_on=(
          "test_track.py::merge_tracks_can_be_created_with_minimal_data",
          "test_track.py::merge_tracks_can_be_created_with_full_data",
    ),
    scope="session",
)
def merge_tracks_can_be_converted_to_track_data():
    merge = Merge(**load_toml("merge_full.toml"))
    track, = TrackData.from_merge(merge)
    assert track.metadata == Tags(title="My Track Title")
    assert track.cover == Url(href="www.example.com/pic1.jpeg")
    assert tuple(track.parts) == (
        AudioData(
            url="www.example1.com/watch?v=",
            start=None,
            end=None,
        ),
        AudioData(
            url="www.example2.com/watch?v=",
            start=timedelta(seconds=1, milliseconds=500),
            end=timedelta(minutes=43, seconds=21),
        )
    )


@test(
    depends_on=(
          "singles_can_be_converted_to_track_data",
          "split_tracks_can_be_converted_to_track_data",
          "playlists_can_be_converted_to_track_data",
          "merge_tracks_can_be_converted_to_track_data",
    )
)
def tracks_can_be_converted_to_track_data():
    single = Single(**load_toml("single_full.toml"))
    split = Split(**load_toml("split_track_full.toml"))
    playlist = Playlist(**load_toml("playlist_full.toml"))
    merge = Merge(**load_toml("merge_full.toml"))
    verify(
        "\n\n".join(
            pformat(track)
            for tracks in (single, split, playlist, merge)
            for track in TrackData.from_track(tracks)
        )
    )


@test(depends_on=("tracks_can_be_converted_to_track_data",))
def albums_can_be_converted_to_track_data():
    album = Album(**load_toml("album_full.toml"))
    tracks = tuple(TrackData.from_album(album))
    verify(pformat(tracks))


@test(
    depends_on=(
          "tracks_can_be_converted_to_track_data",
          "albums_can_be_converted_to_track_data",
    )
)
def media_can_be_converted_to_track_data():
    data = (
        load_toml("single_full.toml"),
        load_toml("split_track_full.toml"),
        load_toml("playlist_full.toml"),
        load_toml("merge_full.toml"),
        load_toml("album_full.toml"),
    )
    tracks = (Media(**media) for media in data)
    tracks = (TrackData.from_media(media) for media in tracks)
    tracks = (track for track_group in tracks for track in track_group)
    tracks = tuple(tracks)
    verify(pformat(tracks))
