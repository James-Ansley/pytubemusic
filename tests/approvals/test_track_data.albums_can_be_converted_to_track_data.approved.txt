(TrackData(metadata=Tags(title='My Track Title'
                         album='My Album Title'
                         track=1),
           cover=File(path=PosixPath('data/pic.jpeg')),
           parts=[AudioData(url='www.example.com/watch?v=',
                            start=datetime.timedelta(seconds=1, microseconds=500000),
                            end=datetime.timedelta(seconds=201))]),
 TrackData(metadata=Tags(title='My First Track Title'
                         album='My Album Title'
                         track=2),
           cover=Url(href='www.example.com/pic2.jpeg'),
           parts=[AudioData(url='www.example.com/watch?v=',
                            start=datetime.timedelta(seconds=1, microseconds=500000),
                            end=datetime.timedelta(seconds=2601))]),
 TrackData(metadata=Tags(title='My Second Track Title'
                         album='My Album Title'
                         track=3),
           cover=Url(href='www.example.com/pic1.jpeg'),
           parts=[AudioData(url='www.example.com/watch?v=',
                            start=datetime.timedelta(seconds=12, microseconds=300000),
                            end=datetime.timedelta(seconds=2601))]),
 TrackData(metadata=Tags(title='My Track Title'
                         album='My Album Title'
                         track=4),
           cover=Url(href='www.example.com/pic2.jpeg'),
           parts=[PlaylistAudioData(url='www.example.com/playlist?list=',
                                    index=0,
                                    start=datetime.timedelta(seconds=1, microseconds=500000),
                                    end=datetime.timedelta(seconds=2601))]),
 TrackData(metadata=Tags(title='My Other Track Title'
                         album='My Album Title'
                         track=5),
           cover=Url(href='www.example.com/pic1.jpeg'),
           parts=[PlaylistAudioData(url='www.example.com/playlist?list=',
                                    index=2,
                                    start=None,
                                    end=datetime.timedelta(seconds=2601))]),
 TrackData(metadata=Tags(title='My Track Title'
                         album='My Album Title'
                         track=6),
           cover=Url(href='www.example.com/pic1.jpeg'),
           parts=[AudioData(url='www.example.com/watch?v=',
                            start=None,
                            end=None),
                  AudioData(url='www.example.com/watch?v=',
                            start=datetime.timedelta(seconds=1, microseconds=500000),
                            end=datetime.timedelta(seconds=2601))]),
 TrackData(metadata=Tags(title='My Track Title'
                         album='My Album Title'
                         track=7),
           cover=Url(href='www.example.com/pic2.jpeg'),
           parts=[PlaylistAudioData(url='www.example.com/playlist?list=',
                                    index=0,
                                    start=datetime.timedelta(seconds=1, microseconds=500000),
                                    end=datetime.timedelta(seconds=2601)),
                  PlaylistAudioData(url='www.example.com/playlist?list=',
                                    index=2,
                                    start=datetime.timedelta(seconds=1),
                                    end=None),
                  PlaylistAudioData(url='www.example.com/playlist?list=',
                                    index=3,
                                    start=None,
                                    end=None)]),
 TrackData(metadata=Tags(title='My Other Track Title'
                         album='My Album Title'
                         track=8),
           cover=Url(href='www.example.com/pic1.jpeg'),
           parts=[PlaylistAudioData(url='www.example.com/playlist?list=',
                                    index=5,
                                    start=None,
                                    end=datetime.timedelta(seconds=2601))]))
